source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -0,0 +1,166 @@
+
+const absl::string_view Filter::HTTP2_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "http inspector: new connection accepted");
+
+  Network::ConnectionSocket& socket = cb.socket();
+
+  absl::string_view transport_protocol = socket.detectedTransportProtocol();
+  if (!transport_protocol.empty() &&
+      transport_protocol != TransportSockets::TransportSocketNames::get().RawBuffer) {
+    ENVOY_LOG(trace, "http inspector: cannot inspect http protocol with transport socket {}",
+              transport_protocol);
+    return Network::FilterStatus::Continue;
+  }
+
+  ASSERT(file_event_ == nullptr);
+
+  file_event_ = cb.dispatcher().createFileEvent(
+      socket.ioHandle().fd(),
+      [this](uint32_t events) {
+        ASSERT(events == Event::FileReadyType::Read);
+        onRead();
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+
+  cb_ = &cb;
+  return Network::FilterStatus::StopIteration;
+}
+
+void Filter::onRead() {
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const Api::SysCallSizeResult result =
+      os_syscalls.recv(cb_->socket().ioHandle().fd(), buf_, Config::MAX_INSPECT_SIZE, MSG_PEEK);
+  ENVOY_LOG(trace, "http inspector: recv: {}", result.rc_);
+
+  if (result.rc_ == -1 && result.errno_ == EAGAIN) {
+    return;
+  } else if (result.rc_ < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  if (static_cast<uint64_t>(result.rc_) > read_) {
+    const uint8_t* data = buf_ + read_;
+    const size_t len = result.rc_ - read_;
+    read_ = result.rc_;
+    parseHttpHeader(absl::string_view(reinterpret_cast<const char*>(data), len));
+  } else {
+    done(false);
+  }
+}
+
+void Filter::parseHttpHeader(absl::string_view data) {
+  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+    ENVOY_LOG(trace, "http inspector: http2 connection preface found");
+    protocol_ = "HTTP/2";
+    done(true);
+  } else {
+    size_t pos = data.find_first_of("\r\n");
+
+    // Cannot find \r or \n
+    if (pos == absl::string_view::npos) {
+      ENVOY_LOG(trace, "http inspector: no request line detected");
+      return done(false);
+    }
+
+    absl::string_view request_line = data.substr(0, pos);
+    std::vector<std::string> fields = absl::StrSplit(request_line, ' ');
+
+    // Method SP Request-URI SP HTTP-Version
+    if (fields.size() != 3) {
+      ENVOY_LOG(trace, "http inspector: invalid http1x request line");
+      return done(false);
+    }
+
+    if (httpProtocols().count(fields[2]) == 0) {
+      ENVOY_LOG(trace, "http inspector: protocol: {} not valid", fields[2]);
+      return done(false);
+    }
+
+    if (httpMethods().count(fields[0]) == 0) {
```

Comment:
I would eliminate this `if` since you're already logging method below, having this `if` here impact performance.

Reply:
Since we're not supporting all available methods, logging this seems quite excessive. I'd remove all method-related code altogether.

Reply:
Code removed.